src/unix/lwt_domain.mli

```diff
@@ -0,0 +1,64 @@
+(* This file is part of Lwt, released under the MIT license. See LICENSE.md for
+   details, or visit https://github.com/ocsigen/lwt/blob/master/LICENSE.md. *)
+
+
+
+(** This module allows to mix multicore parallelism with [Lwt]
+    cooperative threads. It maintains an extensible pool of domains
+    to which you can detach computations. *)
+
+    val detach : ('a -> 'b) -> 'a -> 'b Lwt.t
+    (** [detach f x] runs the computation [f x] in a separate domain in parallel.
+        [detach] evaluates to an Lwt promise, which is pending until the
+        domain completes execution.
+        [detach] calls {!simple_init} internally, which means that the number of
+        domains is capped by default at four. If you would like a
+        higher limit, call {!init} or {!set_bounds} directly.
+        Note that Lwt thread-local storage (i.e., {!Lwt.with_value}) cannot be
+        safely used from within [f]. The same goes for most of the rest of Lwt. If
+        you need to run an Lwt thread in [f], use {!run_in_main}. *)
+
+  val run_in_main : (unit -> 'a Lwt.t) -> 'a
```

Comment:
IIUC, if you run `run_in_main f` from a detached computation, then this computation is essentially frozen until the promise `f ()` has resolved. Is this correct understanding?

What happens to this domain in the mean-time? Can it be used by other detached computations?

Is it possible to make a dead-lock by making a promise that waits for a detached computation to finish but the computation waits on the Lwt promise to resolve?

Is it possible to run out of domains by freezing too many computations waiting for promises?

I think this function is useful, but it should be used sparingly. The documentation (I know it's not in its final form rn, just planning ahead) should have a precise description of what happens (a) to the computation, (b) to the domain that the computation runs on, (c) to the promise that's created. In addition, depending on the answer to some of the questions above, the documentation may also need to include warnings, usage notes, and such.

Reply:
I concur that `run_in_main` needs to be used only sparingly, if at all it is.

>IIUC, if you run run_in_main f from a detached computation, then this computation is essentially frozen until the promise f () has resolved. Is this correct understanding?

Yes that's correct, the calling domain is frozen until the computation is resolved. Other detached computations cannot use that domain since it's blocked, waiting for the promise f () to resolve. This is due to the fact that the function returns the final result and not the promise.

>Is it possible to make a dead-lock by making a promise that waits for a detached computation to finish but the computation waits on the Lwt promise to resolve?

Yes I'd imagine that's possible, and it's not ideal.

>Is it possible to run out of domains by freezing too many computations waiting for promises?

Possible that many domains could be blocked waiting for promises, but I think they'd eventually resolve or max out the queue limit (if we have one).

Updating the documentation is a good idea, will do that!

Reply:
Ok. Essentially I think this problem is one of documentation. `run_in_main` should have a biggish warning about freeze and about the possibility of dead-locks.

When finalising this and getting it merged and all, we can then add to the documentation with an example of how to avoid running `run_in_main` and to prefer a stepping mechanism where the parallel computation returns a partial result (similar to what parsers often do when they need to wait for more input). This part really can wait a bit: it's not worth putting some effort in writing this doc until the interface is finalised.